### ISSUE TITLE
Vault OOM

### ISSUE THREAD

**Issue by tcolgate** [hashicorp/vault#2899]
Hi,
  We are using vault with the etcd backend (currently version 1.7.2). tokem tidy has been run.
  I am building out container secret injection system around vault approles and token roles. We have some rather convoluted logic that creates policies, token roles, and approles on demand. Our vault servers are queried fairly continuously by our puppet sync tooling. Normally this results in a vault working set of about 50MB.
  Since introducing the new dynamic policy/role/approle code, our vault server has been periodically OOM'ing. At this point it has reached about 1.5GB and is killed by our container orchestration.
  I can recreate this fairly reliably by querying token roles. A repeated list, or read, of a token role is enough to trigger the problem. A few dozen reads/lists in succession is enough to trigger the OOM. We discovered this with the tool that creates the approles and policies, this does about 6 vault calls, creating a policy, a token role, and an approle, the retrieving a role-id. Running this two or three times, is enough to kill vault.
  I have  a couple of observations:
- doing the read or list seems to block vault (output in the audit logs pauses between runs of a list or read on approle)
- Just before the OOM we see [update: the actual revocations may be a read herring, seems those errors will have only started today, whereas the problem has been ovserved over the last 3 days since develpement on this flow began]:
```2017/06/21 12:54:09.749727 [ERROR] expire: failed to revoke lease: lease_id=aws/creds/opentaglogs/0c975e5d-c6b5-0448-6750-afe0164d4c66 error=failed to revoke entry: resp:(*logical.Response)(nil) err:NoSuchEntity: The user with name vault--opentaglogs-1472122451-4706 cannot be found.
2017/06/21 12:54:09.750688 [ERROR] expire: failed to revoke lease: lease_id=aws/creds/opentaglogs/f3e626f7-97b5-3a23-b4fa-721e79c3a56e error=failed to revoke entry: resp:(*logical.Response)(nil) err:NoSuchEntity: The user with name vault--opentaglogs-1472121047-1133 cannot be found.
2017/06/21 12:54:09.754228 [ERROR] expire: failed to revoke lease: lease_id=aws/creds/opentaglogs/ea4344e4-a27b-080d-538d-f9d066c90a78 error=failed to revoke entry: resp:(*logical.Response)(nil) err:NoSuchEntity: The user with name vault--opentaglogs-1472121455-7026 cannot be found.
2017/06/21 12:54:09.817231 [ERROR] expire: failed to revoke lease: lease_id=aws/creds/opentaglogs/2d775129-e9a5-f925-8e5a-8f029809d060 error=failed to revoke entry: resp:(*logical.Response)(nil) err:NoSuchEntity: The user with name vault--opentaglogs-1472121036-3304 cannot be found.
Container exitted: error = "", exitcode = 137
Container died due to OOM
```

**Comment by tcolgate** [comment]
It may be relevant that it only seems to happen when using a non-root token (that token has a fairly elaborate policy)

**Comment by jefferai** [comment]
Unfortunately there really isn't enough to go on here to figure out what the issue might be...there's no configuration information, no full logs, no indication of which commands you're running. It would help immensely if you could figure out a repeatable test case with a dev server that we can use to try to track the issue down.

**Comment by tcolgate** [comment]
I've disabled audit, incase our audit backend was the problem.
I've disabled the job that we have sync'ing things to puppet.
The log attached shows the output from the vault servers. I waiting for vault to be up, unsealed and ready, started tailing the logs. Then,manually in a shell repeatedly issue `vault read /auth/token/roles/read-qubit-infra-stg.generic-eu.dubber`, I ran about 40 or so, they gradually slow down, and ten eventually error when the service OOMs.
[log.txt](https://github.com/hashicorp/vault/files/1091812/log.txt)

**Comment by tcolgate** [comment]
Config is straightforward.
```backend "etcd" {
  advertise_addr = "https://vault-staging.XXX.com"
  address = "https://etcd-staging.XXX.com:2379"
  username = "vault"
  password = "XXX"
  tls_key_file = "/vault/config/etcd-client-key.pem"
  tls_cert_file = "/vault/config/etcd-client.pem"
  tls_ca_file = "/vault/config/etcd-ca.pem"
  ha_enabled = "true"
}

listener "tcp" {
  address     = ":10443"
  tls_key_file   = "/vault/config/server-key.pem"
  tls_cert_file  = "/vault/config/server.pem"
}

listener "tcp" {
  address = "127.0.0.1:8200"
  tls_disable = 1
}
```

**Comment by tcolgate** [comment]
constructing a test case is going to take me a while (presuming I can recreate the problem at all). I'm out of the office until next week unfortunately too.

**Comment by tcolgate** [comment]
@jefferai added some logs and out config (mildly sanitized). I appreciate there's still not a lot to go on.

**Comment by jefferai** [comment]
@tcolgate Can you provide the logs of the Vault server please?

**Comment by tcolgate** [comment]
@jefferai the logs are as seen in logs.txt.

**Comment by jefferai** [comment]
Those are audit logs...

**Comment by tcolgate** [comment]
That's the entire output of vault during the requests, there are no other errors, and I don't believe there's anything else on stderr/stdout. I can double check.

**Comment by jefferai** [comment]
@tcolgate The top part are audit logs. Those are not normally sent to stdout, I don't know where those are coming from.

The bottom part is stdout, which has informational values, but Vault's actual server logs go to stderr. Those are what would be helpful to see, ideally with `-log-level=trace`.

**Comment by tcolgate** [comment]
These are the logs for an entire server lifetime. There is non-audit info in there too
[log.txt](https://github.com/hashicorp/vault/files/1092014/log.txt)

**Comment by jefferai** [comment]
I'm really quite confused what's going on here. You said that you disabled auditing, but auditing is being interspersed in your logs, which makes no sense unless auditing *is* enabled and you are sending auditing to stdout. You also said that you're using approle and/or token store roles, but what you're showing here includes a number of failures from AWS.

**Comment by tcolgate** [comment]
I disabled our socket based auditing. File auditing is still on.

**Comment by tcolgate** [comment]
We're using token roles and approle in our new flow, but I can trigger the oom purely by reading the /auth/token/roles/thing endpoint. The AWS errors are from another process that I can't disable (as is some of the other token activity). That activitly has been ongoing for months in prod and has not caused problems.

**Comment by jefferai** [comment]
You can use `sys/revoke-force` to force Vault to drop leases that cannot be properly revoked.

**Comment by tcolgate** [comment]
It appears that I can trigger an OOM hitting that endpoint even with a non
existent token role.

**Comment by jefferai** [comment]
Hi @tcolgate ,

I just ran the following:

`vault write -f auth/token/roles/foo; for i in $(seq 1 10000); do vault read auth/token/roles/foo; done`

I see no increase in memory consumed by Vault during or after these 10000 iterations.

**Comment by Chronojam** [comment]
Hey,

Ive done some testing on this, and i was able to replicate with the following setup:

1 Vault instance, with 3 backing etcd instances:

**Config**

```
backend "etcd" {
  advertise_addr = "http://vault:8200"
  address = "http://example-etcd-cluster-0000:2379,http://example-etcd-cluster-0001:2379,http://example-etcd-cluster-0002:2379"
  ha_enabled = "true"
}

listener "tcp" {
    address = "0.0.0.0:8200"
    tls_disable = 1
}
```

**Steps to reproduce:**



1.  Initialize vault instance, unseal and set your token to the initial root token, enable approle authentication

```
vault init
vault unseal x
vault unseal y
vault unseal z

export VAULT_TOKEN=ROOT_TOKEN

vault auth-enable approle
```
2. Add the following policy, then generate a new token from it. Set your token to that one instead of the root token.
```
$ cat test.hcl
path "sys/policy/test/*" {
  capabilities = ["sudo", "create", "update", "list"]
}

path "auth/approle/role/*" {
  capabilities = ["create", "update", "list", "read"]
}

path "auth/token/roles/*" {
  capabilities = ["create", "update"]

  denied_parameters = {
    allowed_policies = ["root"]
  }
}

path "auth/token/roles/*" {
  capabilities = ["read", "list"]
}

path "auth/token/create/*" {
  capabilities = ["create", "update"]
}

$ vault policy-write test test.policy
$ vault token-create -ttl=80h -policy=test
$ export VAULT_TOKEN=NEW_TOKEN_FROM_ABOVE
```
3. Add the following policies:
```
$ cat foo-read.policy 
path "foo/bar/baz/*" {
        capabilities = ["read", "list"] 
}

$ cat foo-write.policy 
path "foo/bar/baz/*" {
        capabilities = ["create", "update"]
}

$ cat foo-getsecretid.policy 
path "auth/approle/role/foo/secret-id" {
		capabilities = ["create", "update"]
}

$ cat foo-getroleid.policy
path "auth/approle/role/foo/role-id" {
		capabilities = ["read"]
}

$ vault policy-write test/foo-read.policy foo-read.policy
$ vault policy-write test/foo-write.policy foo-write.policy
$ vault policy-write test/foo-getsecretid.policy foo-getsecretid.policy
$ vault policy-write test/foo-getroleid.policy foo-getroleid.policy
```
4. Start creating roles and tokens for those policies a bunch of times.
```
$ cat vault-load.sh 
#!/bin/bash

ROLE_NAME=foo.bar.baz
THINGS="foo-read foo-write foo-getsecretid foo-getroleid"

for i in `seq 1 10`; do
    for pol in $THINGS; do
        echo $pol
        S_POLICY="test/$pol.policy"
        TOK_ROLE="$pol.policy"

        # Get RoleID
        vault write -f auth/approle/role/$ROLE_NAME policies=foo-read.policy secret_id_num_uses=1 secret_id_ttl=5m token_ttl=5m token_max_ttl=5m
        vault read auth/approle/role/$ROLE_NAME/role-id

        # Get SecretID
        vault write -f auth/approle/role/$ROLE_NAME/secret-id

        # Create a token role
        vault write -f auth/token/roles/$TOK_ROLE allowed_policies=$S_POLICY disallowed_policies=root

        # Create a new token for our role.
        vault token-create -policy=$S_POLICY -role=$TOK_ROLE
    done
done

$ ./vault-load.sh
```
This causes the vault instance to rapidly use memory and slow down till it crashes.
It's very important to note that when you run the above, but with the root token instead - it works fine.

Im going to try this with a vault dev server to see if i can replicate the behaviour there.

Hope this is helpful!

**Comment by Chronojam** [comment]
Hm ok,

I can replicate this with vault 0.7.2 on a local dev server, but not with 0.7.3

Ill test again with the full stack on 0.7.3

**Comment by jefferai** [comment]
Based on the description of your test case it may well be related to https://github.com/hashicorp/vault/issues/2804 which was fixed in 0.7.3. Will be interesting to know how your full stack testing goes.

**Comment by tcolgate** [comment]
Upgraded to 0.7.3, we can't trigger the OOM at the moment, all looks good.
Will close after a little further testing.

**Comment by tcolgate** [comment]
I've updated the title for possterity.
Closing. This appears to be resolved in 0.7.3.
